features/environment.py

```diff
@@ -98,36 +97,70 @@ def from_environ(cls) -> "UAClientBehaveConfig":
 
 
 def before_all(context: Context) -> None:
-    """behave will invoke this before anything else happens.
-
-    In this function, we launch a container, install ubuntu-advantage-tools and
-    then capture an image.  This image is then reused by each feature, reducing
-    test execution time.
-    """
+    """behave will invoke this before anything else happens."""
     userdata = context.config.userdata
     context.reuse_container = userdata.get("reuse_container")
+    context.series_image_name = {}
+    context.series_reuse_image = ""
     context.config = UAClientBehaveConfig.from_environ()
-    if context.config.reuse_image is None:
-        create_trusty_uat_lxd_image(context)
-    else:
-        context.image_name = context.config.reuse_image
+    if context.config.reuse_image:
+        series = lxc_get_series(context.config.reuse_image, image=True)
+        if series is not None:
+            context.series_reuse_image = series
```

Comment:
shouldn't we also set context.series_image_name[series] = context.config.reuse_image here too?

Reply:
No. It is in `before_scenario` because it knows which series is currently running, so it avoids, for instance: providing a trusty image and then running focal with that image (in case you do not provide --tags=series.trusty)

Reply:
Sorry, I must be missing something here. reuse_image is provided and will be either a focal or trusty image. Your lxc_get_series properly determines what series that reuse_image is running. If it is `trusty`, we only need to set this once context.series_image_name["trusty"] = contest.config.reuse_image.  We don't need to do this every before_scenario right because it was already set for trusty, regardless of whether the effective tag on a series is trusty or focal or xenial etc right?

Reply:
And if you are setting up series_image_name properly for the proper series that  reuse_image that was provided, then in your before_scenario you can simplify the logic to:

```
if series not in context.series_image_name:
    create_uat_lxd_image(context, series)
```

Reply:
Yes. Sorry. I thought you'd like to move all series info to before_all. I've changed it. Thanks